File: its-my-training-day/training-grounds/leetcode-top-75/questions/graphs/topo-sort/alien-dictionary/solution.ts
import {
  ArrayList,
  Stack,
  Queue,
  ListNode,
  TreeNode,
  MinHeap,
  MaxHeap,
} from "../../../../shared/typescript/ds";

export const alienDictionary = (words: string[]): string => {
  const graph = new Map<string, ArrayList<string>>();
  const indegree = new Map<string, number>();
  for (let i = 0; i < words.length; i += 1) {
    for (let j = 0; j < words[i].length; j += 1) {
      const ch = words[i][j];
      if (!graph.has(ch)) graph.set(ch, new ArrayList<string>());
      if (!indegree.has(ch)) indegree.set(ch, 0);
    }
  }
  for (let i = 0; i + 1 < words.length; i += 1) {
    const a = words[i];
    const b = words[i + 1];
    let j = 0;
    while (j < a.length && j < b.length && a[j] === b[j]) j += 1;
    if (j < a.length && j < b.length) {
      const from = a[j];
      const to = b[j];
      (graph.get(from) as ArrayList<string>).add(to);
      indegree.set(to, (indegree.get(to) as number) + 1);
    } else if (a.length > b.length) {
      return "";
    }
  }
  const queue = new Queue<string>();
  indegree.forEach((value, key) => {
    if (value === 0) queue.enqueue(key);
  });
  const order = new ArrayList<string>();
  while (!queue.isEmpty()) {
    const ch = queue.dequeue() as string;
    order.add(ch);
    const neighbors = graph.get(ch) as ArrayList<string>;
    for (let i = 0; i < neighbors.size(); i += 1) {
      const nxt = neighbors.get(i);
      indegree.set(nxt, (indegree.get(nxt) as number) - 1);
      if (indegree.get(nxt) === 0) queue.enqueue(nxt);
    }
  }
  if (order.size() !== indegree.size) return "";
  return toArray(order).join("");
};

File: its-my-training-day/training-grounds/leetcode-top-75/questions/graphs/topo-sort/alien-dictionary/solution.cpp
#include "../../../../shared/cpp/ds.hpp"

#include <algorithm>
#include <cctype>
#include <cstdint>
#include <functional>
#include <string>
#include <unordered_map>
#include <vector>

std::string alienDictionary(const std::vector<std::string>& words) {
    std::unordered_map<char, ArrayList<char>> graph;
    std::unordered_map<char, int> indegree;
    for (const std::string& w : words) {
        for (char ch : w) {
            if (graph.find(ch) == graph.end()) graph[ch] = ArrayList<char>();
            if (indegree.find(ch) == indegree.end()) indegree[ch] = 0;
        }
    }
    for (int i = 0; i + 1 < static_cast<int>(words.size()); ++i) {
        const std::string& a = words[i];
        const std::string& b = words[i + 1];
        int j = 0;
        while (j < static_cast<int>(a.size()) && j < static_cast<int>(b.size()) && a[j] == b[j]) {
            j += 1;
        }
        if (j < static_cast<int>(a.size()) && j < static_cast<int>(b.size())) {
            graph[a[j]].add(b[j]);
            indegree[b[j]] += 1;
        } else if (a.size() > b.size()) {
            return "";
        }
    }
    Queue<char> queue;
    for (auto& kv : indegree) {
        if (kv.second == 0) queue.enqueue(kv.first);
    }
    ArrayList<char> order;
    while (!queue.empty()) {
        char ch = queue.dequeue();
        order.add(ch);
        ArrayList<char>& neighbors = graph[ch];
        for (std::size_t i = 0; i < neighbors.size(); ++i) {
            char nxt = neighbors.get(i);
            indegree[nxt] -= 1;
            if (indegree[nxt] == 0) queue.enqueue(nxt);
        }
    }
    if (order.size() != indegree.size()) return "";
    std::string result;
    std::vector<char> orderVec = toVector(order);
    result.assign(orderVec.begin(), orderVec.end());
    return result;
}
